exchanges/gateio/gateio_websocket.go

```diff
@@ -764,6 +733,146 @@ func singleSymbolChannel(name string) bool {
 	return false
 }
 
+// ValidateSubscriptions implements subscription.SubscriptionsValidator interface
+func (g *Gateio) ValidateSubscriptions(l subscription.List) error {
+	orderbookGuard := map[key.PairAsset]string{}
+	for _, s := range l {
+		n := channelName(s)
+		if !isRestrictedOrderbookChannel(n) {
+			continue
+		}
+		for _, p := range s.Pairs {
+			k := key.PairAsset{Base: p.Base.Item, Quote: p.Quote.Item, Asset: s.Asset}
+			loadedChanName, ok := orderbookGuard[k]
+			if !ok {
+				orderbookGuard[k] = n
+				continue
+			}
+			if loadedChanName != n {
+				return fmt.Errorf("%w for `%s %s` between `%s` and `%s`, please enable only one type", subscription.ErrExclusiveSubscription, k.Pair(), k.Asset, loadedChanName, n)
+			}
+		}
+	}
+	return nil
+}
+
+// isRestrictedOrderbookChannel returns if the channel is restricted to a single orderbook channel, as multiple orderbook
+// subscriptions enabled will ruin the stored orderbook.
+func isRestrictedOrderbookChannel(name string) bool {
+	switch name {
+	case spotOrderbookUpdateChannel,
+		spotOrderbookChannel,
+		spotOrderbookTickerChannel,
+		futuresOrderbookChannel,
+		futuresOrderbookTickerChannel,
+		futuresOrderbookUpdateChannel,
+		optionsOrderbookChannel,
+		optionsOrderbookTickerChannel,
+		optionsOrderbookUpdateChannel:
+		return true
+	}
+	return false
+}
+
+// ValidateSubscription implements subscription.SubscriptionValidator interface
+func (g *Gateio) ValidateSubscription(s *subscription.Subscription) error {
```

Comment:
Whilst I can see why `ValidateSubscriptions` is needed, this function is just doing what other exchanges do inside the templates.
The template should call a func for interval and level fields, and it should panic on invalid values.
So that'd mean that all Interval logic would be in one template func, and all levels in the other.

Am I missing something?

That leaves just rejecting attempts to subscribe to Margin at all.
Are we saying that websockets are enabled for them but subscriptions are not?
Cos otherwise, we should just set the asset to not websocket supported 🙂

Reply:
[fixed for templates](https://github.com/thrasher-corp/gocryptotrader/pull/1863/commits/d8cca4e2b5db4ef3800c7e285e3ee56772a8a070)

Margin and cross margin are just skipped and return nil so I don't think they should be an issue, they share the same books with spot but I don't know how its going to look once all assets are updated to the new template. Would prefer to leave them for now as it seems out of scope.